qiita_pet/handlers/api_proxy/prep_template.py

```diff
@@ -6,52 +6,383 @@
+
+    Parameters
+    ----------
+    inv_type : str
+        The investigation type
+    user_def_type : str
+        The user-defined investigation type
+    new_type : str
+        The new user-defined investigation_type
+
+    Returns
+    -------
+    str
+        The investigation type chosen by the user
+    """
+    if inv_type == '':
+        inv_type = None
+    elif inv_type == 'Other' and user_def_type == 'New Type':
+        # This is a new user defined investigation type so store it
+        inv_type = new_type
+        ontology = Ontology(convert_to_id('ENA', 'ontology'))
+        ontology.add_user_defined_term(inv_type)
+    elif inv_type == 'Other' and user_def_type != 'New Type':
+        inv_type = user_def_type
+    return inv_type
+
+
+def prep_template_get_req(prep_id, user_id):
+    """Gets the json of the full prep template
+
+    Parameters
+    ----------
+    prep_id : int
+        PrepTemplate id to get info for
+    user_id : str
+        User requesting the sample template info
+
+    Returns
+    -------
+    dict of objects
+    {'status': status,
+     'message': message,
+     'template': {sample: {column: value, ...}, ...}
+    """
+    prep = PrepTemplate(int(prep_id))
+    access_error = check_access(prep.study_id, user_id)
+    if access_error:
+        return access_error
+    df = prep.to_dataframe()
+    return {'status': 'success',
+            'message': '',
+            'template': df.to_dict(orient='index')}
+
+
+def prep_template_summary_get_req(prep_id, user_id):
+    """Get the summarized prep template data for each metadata column
+
+    Parameters
+    ----------
+    prep_id : int
+        PrepTemplate id to get info for
+    user_id : str
+        User requesting the sample template info
+
+    Returns
+    -------
+    dict of objects
+        Dictionary object where the keys are the metadata categories
+        and the values are list of tuples. Each tuple is an observed value in
+        the category and the number of times its seen.
+        Format {'status': status,
+                'message': message,
+                'num_samples': value,
+                'category': [(val1, count1), (val2, count2), ...], ...}
+    """
+    prep = PrepTemplate(int(prep_id))
+    access_error = check_access(prep.study_id, user_id)
+    if access_error:
+        return access_error
+    df = prep.to_dataframe()
+    out = {'num_samples': df.shape[0],
+           'summary': {}}
+
+    # drop the prep_id column if it exists
+    if 'study_id' in df.columns:
+        df.drop('study_id', axis=1, inplace=True)
+    cols = list(df.columns)
+    for column in cols:
+        counts = df[column].value_counts()
+        out['summary'][str(column)] = [(str(key), counts[key])
+                                       for key in natsorted(counts.index)]
+    # Add expected messaging and status info
+    out.update({'status': 'success', 'message': ''})
```

Comment:
Why don't you add those keywords/values at the creation time of the dictionary? I don't see the usage of adding them now and it will make the code more readable.